### PR TITLE
Add Plot sell- and buying

### PR DESCRIPTION
**Pull request by DerDevHD**
## Introduction
Hi. Y'all waited for Plot selling and buying. And now its here!

### Relevant issues
/

### API changes
There are added 2 new Methods for the MyPlot class:
```php
MyPlot::sellPlot(Plot $plot, float $price)
MyPlot::buyPlot(Plot $plot, Player $player)
```

### Behavioural changes
/

## Backwards compatibility
It's 100% backwards compatible

## Follow-up
Improve the phpdoc plz
Improve the english translations plz
Maybe test the MySQL provider

## Tests
<!--
Tested it successfully without any errors. 
MySQL Database isnt tested yet, but should work.

### CODE REVIEW

File: src/MyPlot/provider/MySQLProvider.php [hunk 1]
```diff
@@ -48,9 +48,10 @@ public function __construct(MyPlot $plugin, int $cacheSize = 0, array $settings
 		$this->db = new \mysqli($settings['Host'], $settings['Username'], $settings['Password'], $settings['DatabaseName'], $settings['Port']);
 		if($this->db->connect_error)
 			throw new \RuntimeException("Failed to connect to the MySQL database: " . $this->db->connect_error);
-		$this->db->query("CREATE TABLE IF NOT EXISTS plots (id INT PRIMARY KEY AUTO_INCREMENT, level TEXT, X INT, Z INT, name TEXT, owner TEXT, helpers TEXT, denied TEXT, biome TEXT, pvp INT);");
+		$this->db->query("CREATE TABLE IF NOT EXISTS plots (id INT PRIMARY KEY AUTO_INCREMENT, level TEXT, X INT, Z INT, name TEXT, owner TEXT, helpers TEXT, denied TEXT, biome TEXT, pvp INT, price FLOAT);");
 		try{
 			$this->db->query("ALTER TABLE plots ADD COLUMN pvp INT AFTER biome;");
+			$this->db->query("ALTER TABLE plots ADD COLUMN price FLOAT AFTER pvp;");
```
Comment (jasonwynn10) [inline src/MyPlot/provider/MySQLProvider.php:54]: this needs its own try/catch in case the line above crashes

**Comment by MonoAdrian23** [comment]
We all need merge, not Plot sell and buy

File: src/MyPlot/subcommand/BuySubCommand.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+	/**
+	 * @param Player $sender
+	 * @param string[] $args
+	 *
+	 * @return bool
+	 */
+	public function execute(CommandSender $sender, array $args) : bool {
+		if($this->getPlugin()->getEconomyProvider() === null){
+			$sender->sendMessage("NO-ECONOMY-PLUGIN");
```
Comment (jasonwynn10) [inline src/MyPlot/subcommand/BuySubCommand.php:33]: uhmm what's this?

File: src/MyPlot/subcommand/SellSubCommand.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+	/**
+	 * @param Player $sender
+	 * @param string[] $args
+	 *
+	 * @return bool
+	 */
+	public function execute(CommandSender $sender, array $args) : bool {
+		if($this->getPlugin()->getEconomyProvider() === null){
+			$sender->sendMessage("NO-ECONOMY-PLUGIN");
```
Comment (jasonwynn10) [inline src/MyPlot/subcommand/SellSubCommand.php:32]: and again, what's this?

File: src/MyPlot/subcommand/SellSubCommand.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+			return true;
+		}
+		if($plot->isForSale()){
+			$sender->sendMessage(TextFormat::RED . $this->translateString("sell.already"));
+			return true;
+		}
+		if(!is_numeric($price = $args[0]))
+			return false;
+		if((float) $price <= 0){
+			$sender->sendMessage(TextFormat::RED . $this->translateString("sell.minprice"));
```
Comment (jasonwynn10) [inline src/MyPlot/subcommand/SellSubCommand.php:57]: min price is assuming there will be a max price, but auctioning will have an unlimited max price

**Review by jasonwynn10 — CHANGES_REQUESTED** [review]

File: resources/eng.ini [hunk 5]
```diff
@@ -291,4 +292,24 @@ kick.attemptkick=Player {%0} attempt to kick you
...
 kick.form=Kick
-kick.dropdown=Player Name
\ No newline at end of file
+kick.dropdown=Player Name
+
+; the /p sell Subcommand
+sell.name=sell
+sell.desc=Sells a plot
+sell.usage=/p sell <price: float>
+sell.unclaimed=This plot isn't claimed
```
[the remark this answers is not in the document]
Comment (DerDevHD) [reply resources/eng.ini]: fixed

File: resources/eng.ini [hunk 6]
```diff
@@ -291,4 +292,24 @@ kick.attemptkick=Player {%0} attempt to kick you
...
+sell.already=This plot is already for sale
+sell.minprice=The price should be greater than 0
+sell.success=You successfully released your plot for selling (${%0})
+
+; the /p buy Subcommand
+buy.name=buy
+buy.desc=Buys a plot
+buy.usage=/p buy
+buy.noself=You can't buy your plot from yourself
```
[the remark this answers is not in the document]
Comment (DerDevHD) [reply resources/eng.ini]: fixed

File: src/MyPlot/provider/MySQLProvider.php (hunk 1 above)
Comment (DerDevHD) [reply src/MyPlot/provider/MySQLProvider.php:54]: okay

File: src/MyPlot/subcommand/BuySubCommand.php (hunk 2 above)
Comment (DerDevHD) [reply src/MyPlot/subcommand/BuySubCommand.php:33]: sorry, forgot that

File: src/MyPlot/subcommand/SellSubCommand.php (hunk 3 above)
Comment (DerDevHD) [reply src/MyPlot/subcommand/SellSubCommand.php:32]: forgot that...

File: src/MyPlot/subcommand/SellSubCommand.php (hunk 4 above)
Comment (DerDevHD) [reply src/MyPlot/subcommand/SellSubCommand.php:57]: okay, I'll add this

**Review by DerDevHD — COMMENTED** [review]
> We all need merge, not Plot sell and buy

shut tf off

File: src/MyPlot/subcommand/SellSubCommand.php (hunk 4 above)
Comment (DerDevHD) [reply src/MyPlot/subcommand/SellSubCommand.php:57]: Should it be a config value or a permission node?

File: resources/langtemplate.ini [hunk 7]
```diff
@@ -4,6 +4,7 @@
 
 ; Popup when entering a plot
 popup=
+popup.buy
```
Comment (jasonwynn10) [inline resources/langtemplate.ini:7]: missed an `=`

File: src/MyPlot/provider/YAMLDataProvider.php [hunk 8]
```diff
@@ -201,6 +205,7 @@ public function getNextFreePlot(string $levelName, int $limitXZ = 0) : ?plot {
 	}
 
 	public function close() : void {
+		$this->yaml->save(); // why don't save it on close??
```
Comment (jasonwynn10) [inline src/MyPlot/provider/YAMLDataProvider.php:208]: this is outside the scope of the PR

File: src/MyPlot/provider/YAMLDataProvider.php [hunk 9]
```diff
@@ -156,7 +160,7 @@ public function getPlotsByOwner(string $owner, string $levelName = "") : array {
 	 *
 	 * @return Plot|null
 	 */
-	public function getNextFreePlot(string $levelName, int $limitXZ = 0) : ?plot {
+	public function getNextFreePlot(string $levelName, int $limitXZ = 0) : ?Plot {
```
Comment (jasonwynn10) [inline src/MyPlot/provider/YAMLDataProvider.php:163]: this is outside the scope of the PR

File: src/MyPlot/provider/YAMLDataProvider.php [hunk 10]
```diff
@@ -115,6 +116,7 @@ public function getPlotsByOwner(string $owner, string $levelName = "") : array {
 		if($levelName != "") {
 			/** @var int[] $levelKeys */
 			$levelKeys = array_keys($plots, $levelName);
+			/** @var int[] $ownerKeys */
```
Comment (jasonwynn10) [inline src/MyPlot/provider/YAMLDataProvider.php:119]: this is outside the scope of the PR

File: src/MyPlot/provider/JSONDataProvider.php [hunk 11]
```diff
@@ -32,13 +32,14 @@ public function __construct(MyPlot $plugin, int $cacheSize = 0) {
...
-			$plots[$plot->id] = ["level" => $plot->levelName, "x" => $plot->X, "z" => $plot->Z, "name" => $plot->name, "owner" => $plot->owner, "helpers" => $plot->helpers, "denied" => $plot->denied, "biome" => $plot->biome, "pvp" => $plot->pvp];
+			$plots[$plot->id] = ["level" => $plot->levelName, "x" => $plot->X, "z" => $plot->Z, "name" => $plot->name, "owner" => $plot->owner, "helpers" => $plot->helpers, "denied" => $plot->denied, "biome" => $plot->biome, "pvp" => $plot->pvp, "price" => $plot->price];
 		}else{
 			$id = $this->json->get("count", 0) + 1;
 			$plot->id = $id;
-			$plots[$id] = ["level" => $plot->levelName, "x" => $plot->X, "z" => $plot->Z, "name" => $plot->name, "owner" => $plot->owner, "helpers" => $plot->helpers, "denied" => $plot->denied, "biome" => $plot->biome, "pvp" => $plot->pvp];
+			$plots[$id] = ["level" => $plot->levelName, "x" => $plot->X, "z" => $plot->Z, "name" => $plot->name, "owner" => $plot->owner, "helpers" => $plot->helpers, "denied" => $plot->denied, "biome" => $plot->biome, "pvp" => $plot->pvp, "price" => $plot->price];
 			$this->json->set("count", $id);
 		}
+		$this->json->set("plots", $plots); // i think u forgot that jason?
```
Comment (jasonwynn10) [inline src/MyPlot/provider/JSONDataProvider.php:42]: this is outside the scope of the PR

**Review by jasonwynn10 — CHANGES_REQUESTED** [review]

File: src/MyPlot/provider/JSONDataProvider.php (hunk 11 above)
Comment (DerDevHD) [reply src/MyPlot/provider/JSONDataProvider.php:42]: I know... Should I rebase the changes?

Comment (jasonwynn10) [reply src/MyPlot/provider/JSONDataProvider.php:42]: up to you. I'll probably commit it to master before this PR gets merged

Comment (DerDevHD) [reply src/MyPlot/provider/JSONDataProvider.php:42]: Okay, you'll change it and push it to master, and I'll rebase.

File: resources/eng.ini [hunk 12]
```diff
@@ -291,4 +293,24 @@ kick.attemptkick=Player {%0} attempt to kick you
...
+sell.minprice=The price should be greater than 0
+sell.success=You successfully released your plot for selling (${%0})
+
+; the /p buy Subcommand
+buy.name=buy
+buy.desc=Buys a plot
+buy.usage=/p buy
+buy.noself="You can't buy your plot from yourself"
+buy.notinsale="This plot isn't in sale"
+buy.nomoney=You don't have enough money for buying this plot
```
Comment (jasonwynn10) [inline resources/eng.ini:313]: another line that needs quotes.

**Review by jasonwynn10 — CHANGES_REQUESTED** [review]

**Comment by DerDevHD** [comment]
I'm happy about it; What do you think Jason?

**Comment by jasonwynn10** [comment]
Test and Update the German translations please
